compiler/rustc_feature/src/unstable.rs

```diff
@@ -649,6 +651,59 @@ declare_features! (
     // -------------------------------------------------------------------------
 );
 
+impl Features {
+    pub fn dump_feature_usage_metrics(&self, metrics_dir: &Path) {
+        #[derive(serde::Serialize)]
+        struct LibFeature {
+            symbol: String,
+        }
+
+        #[derive(serde::Serialize)]
+        struct LangFeature {
+            symbol: String,
+            since: Option<String>,
+        }
+
+        #[derive(serde::Serialize)]
+        struct FeatureUsage {
+            lib_features: Vec<LibFeature>,
+            lang_features: Vec<LangFeature>,
+        }
+
+        // TODO (DECIDE): How fine grained do we want to track feature usage?
```

Comment:
use crate name + crate-sha, will handle anonymization in any upload process we make available.